File: binding/src/network/libqhostinfo.cpp
#include <QHostInfo>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qhostinfo.h>
#include "libqhostinfo.h"
#include "libqhostinfo.hxx"

QHostInfo* QHostInfo_new() {
	 return new QHostInfo();
}

QHostInfo* QHostInfo_new2(const QHostInfo* d) {
	 return new QHostInfo(*d);
}

QHostInfo* QHostInfo_new3(int lookupId) {
	 return new QHostInfo(lookupId);
}

void QHostInfo_OperatorAssign(QHostInfo* self, const QHostInfo* d) {
	self->operator=(*d);
}

void QHostInfo_Swap(QHostInfo* self, QHostInfo* other) {
	self->swap(*other);
}

libqt_string QHostInfo_HostName(const QHostInfo* self) {
	QString _ret = self->hostName();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QHostInfo_SetHostName(QHostInfo* self, const libqt_string name) {
	self->setHostName(QString::fromUtf8(name.data, name.len));
}

libqt_list QHostInfo_Addresses(const QHostInfo* self) {
	return self->addresses();
}

void QHostInfo_SetAddresses(QHostInfo* self, const libqt_list addresses) {
	self->setAddresses(*addresses);
}

int QHostInfo_Error(const QHostInfo* self) {
	return self->error();
}

void QHostInfo_SetError(QHostInfo* self, int errorVal) {
	self->setError(static_cast<QHostInfo::HostInfoError>(errorVal));
}

libqt_string QHostInfo_ErrorString(const QHostInfo* self) {
	QString _ret = self->errorString();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QHostInfo_SetErrorString(QHostInfo* self, const libqt_string errorString) {
	self->setErrorString(QString::fromUtf8(errorString.data, errorString.len));
}

void QHostInfo_SetLookupId(QHostInfo* self, int id) {
	self->setLookupId(id);
}

int QHostInfo_LookupId(const QHostInfo* self) {
	return self->lookupId();
}

int QHostInfo_LookupHost(const libqt_string name, const QObject* receiver, const char* member) {
	return QHostInfo::lookupHost(QString::fromUtf8(name.data, name.len), receiver, member);
}

void QHostInfo_AbortHostLookup(int lookupId) {
	QHostInfo::abortHostLookup(lookupId);
}

QHostInfo* QHostInfo_FromName(const libqt_string name) {
	return new QHostInfo(QHostInfo::fromName(QString::fromUtf8(name.data, name.len)));
}

libqt_string QHostInfo_LocalHostName() {
	QString _ret = QHostInfo::localHostName();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QHostInfo_LocalDomainName() {
	QString _ret = QHostInfo::localDomainName();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QHostInfo_Delete(QHostInfo* self) {
    delete self;
}

File: binding/src/network/libqhostinfo.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qhostinfo.html
pub const qhostinfo = struct {

    /// New constructs a new QHostInfo object.
    pub fn New() QtC.QHostInfo {
        return qtc.QHostInfo_new();
    }


    /// New2 constructs a new QHostInfo object.
    pub fn New2(d: ?*anyopaque) QtC.QHostInfo {
        return qtc.QHostInfo_new2(@ptrCast(d));
    }


    /// New3 constructs a new QHostInfo object.
    pub fn New3(lookupId: i32) QtC.QHostInfo {
        return qtc.QHostInfo_new3(lookupId);
    }


    pub fn OperatorAssign(self: ?*anyopaque, d: ?*anyopaque) void {
        qtc.QHostInfo_OperatorAssign(@ptrCast(self), @ptrCast(d));
    }

    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QHostInfo_Swap(@ptrCast(self), @ptrCast(other));
    }

    pub fn HostName(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QHostInfo_HostName(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qhostinfo.HostName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetHostName(self: ?*anyopaque, name: []const u8) void {
        const name_str = qtc.libqt_string{
    .len = name.len,
    .data = name.ptr,
};
qtc.QHostInfo_SetHostName(@ptrCast(self), name_str);
    }

    pub fn Addresses(self: ?*anyopaque, ) []const u8 {
        return qtc.QHostInfo_Addresses(@ptrCast(self));
    }

    pub fn SetAddresses(self: ?*anyopaque, addresses: []const u8) void {
        qtc.QHostInfo_SetAddresses(@ptrCast(self), @ptrCast(addresses));
    }

    pub fn Error(self: ?*anyopaque, ) i32 {
        return qtc.QHostInfo_Error(@ptrCast(self));
    }

    pub fn SetError(self: ?*anyopaque, errorVal: i32) void {
        qtc.QHostInfo_SetError(@ptrCast(self), @intCast(errorVal));
    }

    pub fn ErrorString(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QHostInfo_ErrorString(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qhostinfo.ErrorString: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetErrorString(self: ?*anyopaque, errorString: []const u8) void {
        const errorString_str = qtc.libqt_string{
    .len = errorString.len,
    .data = errorString.ptr,
};
qtc.QHostInfo_SetErrorString(@ptrCast(self), errorString_str);
    }

    pub fn SetLookupId(self: ?*anyopaque, id: i32) void {
        qtc.QHostInfo_SetLookupId(@ptrCast(self), id);
    }

    pub fn LookupId(self: ?*anyopaque, ) i32 {
        return qtc.QHostInfo_LookupId(@ptrCast(self));
    }

    pub fn LookupHost(name: []const u8, receiver: ?*anyopaque, member: []const u8) i32 {
        const name_str = qtc.libqt_string{
    .len = name.len,
    .data = name.ptr,
};
return qtc.QHostInfo_LookupHost(name_str, @ptrCast(receiver), @ptrCast(member));
    }

    pub fn AbortHostLookup(lookupId: i32) void {
        qtc.QHostInfo_AbortHostLookup(lookupId);
    }

    pub fn FromName(name: []const u8) QtC.QHostInfo {
        const name_str = qtc.libqt_string{
    .len = name.len,
    .data = name.ptr,
};
return qtc.QHostInfo_FromName(name_str);
    }

    pub fn LocalHostName(, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QHostInfo_LocalHostName();
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qhostinfo.LocalHostName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn LocalDomainName(, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QHostInfo_LocalDomainName();
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qhostinfo.LocalDomainName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QHostInfo_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/hostinfo.html#types
pub const enums = struct {
    pub const HostInfoError = enum {
        pub const NoError: i32 = 0;
        pub const HostNotFound: i32 = 1;
        pub const UnknownError: i32 = 2;
    };

};
